kafka/metrics.go

```diff
@@ -0,0 +1,103 @@
+package kafka
+
+import (
+	"context"
+	"runtime"
+	"time"
+
+	"github.com/twmb/franz-go/pkg/kgo"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/metric"
+)
+
+const instrumentName = "github.com/elastic/apm-queue/kafka"
+const unitCount = "1"
+
+type instruments struct {
+	MessageBuffered metric.Int64Counter
+	MessageProduced metric.Int64Counter
+	WriteErrors     metric.Int64Counter
+}
+
+type kgoHooks struct {
+	instruments instruments
+}
+
+func NewKgoHooks(mp metric.MeterProvider) *kgoHooks {
+	m := mp.Meter(instrumentName)
+
+	a, err := m.Int64Counter(
+		"message.produced",
+		metric.WithDescription("The total number of message produced"),
+		metric.WithUnit(unitCount),
+	)
+	if err != nil {
+		panic(err)
+	}
+
+	b, err := m.Int64Counter(
+		"write.error.count",
+		metric.WithDescription("The total number of error occurred on write"),
+		metric.WithUnit(unitCount),
+	)
+	if err != nil {
+		panic(err)
+	}
+
+	c, err := m.Int64Counter(
+		"message.buffered",
+		metric.WithDescription("The total number of message buffered from produce batches"),
+		metric.WithUnit(unitCount),
+	)
+	if err != nil {
+		panic(err)
+	}
+
+	return &kgoHooks{
+		instruments{
+			MessageBuffered: c,
+			MessageProduced: a,
+			WriteErrors:     b,
+		},
+	}
+}
+
+// https://pkg.go.dev/github.com/twmb/franz-go/pkg/kgo#HookProduceBatchWritten
+func (h *kgoHooks) OnProduceBatchWritten(meta kgo.BrokerMetadata, topic string, partition int32, metrics kgo.ProduceBatchMetrics) {
```

Comment:
Shouldn't this be `OnProduceRecordUnbuffered`? Otherwise any bytes written to the broker will increment the counter. It is implemented by the `kotel.Tracer` to end a span, but in this case, we could use it to record that a record has been produced.

https://pkg.go.dev/github.com/twmb/franz-go/pkg/kgo#HookProduceRecordUnbuffered

https://github.com/twmb/franz-go/blob/8406043eb1a3eb5e117336b373106a5a9003c9f6/plugin/kotel/tracer.go#L184-L199

Reply:
Thanks for the hint, updated